skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+    can measure the likelihood that a pair of aligned characters are homologous
+    rather than by chance.
+
+    This class provides a generalized interface for substitution matrices. The
+    alphabet usually consists of individual characters, such as nucleotides or
+    amino acids, but it can be generalized to any iterable of scalars (numbers,
+    strings, etc.). Therefore, you may use this class to construct substitution
+    matrices of complicated biological units (such as codons or non-canonical
+    amino acids). The score matrix may be symmetric, as many existing matrices
+    are, or assymetric, where the score of one character substituted by another
+    is unequal to the the other way around. Only square matrices (i.e., numbers
+    of rows and columns are equal) are supported.
+
+    Multiple commonly used nucleotide and amino acid substitution matrices are
+    pre-defined and can be referred to by name. Examples include NUC.4.4 for
+    nucleotides, and variants of BLOSUM and PAM matrices for amino acids.
+
+    ``SubstitutionMatrix`` is a subclass of ``DissimilarityMatrix``. Therefore,
+    all attributes and methods of the latter also apply to the former.
+
+    Examples
+    --------
+    >>> from skbio import SubstitutionMatrix
+    >>> mat = SubstitutionMatrix('ACGT', np.array([
+    ...     [2, -1, -1, -1],
+    ...     [-1, 2, -1, -1],
+    ...     [-1, -1, 2, -1],
+    ...     [-1, -1, -1, 2]]))
+    >>> mat.alphabet
+    ('A', 'C', 'G', 'T')
+    >>> mat.scores
+    array([[ 2., -1., -1., -1.],
+           [-1.,  2., -1., -1.],
+           [-1., -1.,  2., -1.],
+           [-1., -1., -1.,  2.]])
+    >>> mat['A', 'T']
+    -1.0
+    >>> mat['G', 'G']
+    2.0
+    >>> blosum62 = SubstitutionMatrix.by_name('BLOSUM62')
+
+    """
+
+    @property
+    @experimental(as_of='0.5.10')
+    def alphabet(self):
+        """Alphabet of the substitution matrix.
+
+        Each element (character) corresponds to one row/column in the matrix.
+
+        Returns
+        -------
+        tuple
+            Alphabet of the substitution matrix.
+
+        Notes
+        -----
+        This is an alias of ``ids``.
+
+        """
+        return self._ids
+
+    @property
+    @experimental(as_of='0.5.10')
+    def scores(self):
+        """Matrix of substitution scores.
+
+        Each value corresponds to the score of substituting the row character
+        with the column character.
+
+        Returns
+        -------
+        2D np.ndarray
+            Matrix of substitution scores.
+
+        Notes
+        -----
+        This is an alias of ``data``.
+
+        """
+        return self._data
+
+    @experimental(as_of='0.5.10')
+    def __init__(self, alphabet, scores, **kwargs):
+        """Initialize a substitution matrix object with an alphabet and a score
+        matrix.
+        """
+        super().__init__(scores, alphabet, **kwargs)
+
+    @experimental(as_of='0.5.10')
+    def to_dict(self):
```

Comment:
I somewhat expected this to return `{(row, col): value}` rather than 2D? No strong opinion though

Reply:
The current behavior of `to_dict` is consistent with the original approach in skbio (although I think it is bit redundant as one can simply do `mat['M', 'K']` to access a value). Besides, I tend to think that using tuples as dictionary keys, despite valid (as tuples are hashables), may cause confusions to Python beginners.

Reply:
Perhaps though nested `dict`s are also quite complex. As noted though, no strong opinion